source/s3_default_buffer_pool.c

```diff
@@ -0,0 +1,612 @@
+ * Secondary storage delegates directly to system allocator.
+ *
+ * One complication is "forced" buffers. A forced buffer is one that
+ * comes from primary or secondary storage as usual, but it is allowed to exceed
+ * the memory limit. This is only used when we want to use memory from
+ * the pool, but waiting for a normal ticket reservation could cause deadlock.
+ */
+
+struct aws_s3_default_buffer_pool;
+
+struct aws_s3_default_buffer_ticket {
+    size_t size;
+    uint8_t *ptr;
+    size_t chunks_used;
+    bool forced;
+    struct aws_s3_buffer_pool *pool;
+};
+
+/* Default size for blocks array. Note: this is just for meta info, blocks
+ * themselves are not preallocated. */
+static size_t s_block_list_initial_capacity = 5;
+
+/* Amount of mem reserved for use outside of buffer pool.
+ * This is an optimistic upper bound on mem used as we dont track it.
+ * Covers both usage outside of pool, i.e. all allocations done as part of s3
+ * client as well as any allocations overruns due to memory waste in the pool. */
+static const size_t s_buffer_pool_reserved_mem = MB_TO_BYTES(128);
+
+/*
+ * How many chunks make up a block in primary storage.
+ */
+static const size_t s_chunks_per_block = 16;
+
+/*
+ * Max size of chunks in primary.
+ * Effectively if client part size is above the following number, primary
+ * storage along with buffer reuse is disabled and all buffers are allocated
+ * directly using allocator.
+ */
+static const size_t s_max_chunk_size_for_buffer_reuse = MB_TO_BYTES(64);
+
+/* Forced buffers only count against the memory limit up to a certain percent.
+ * For example: if mem_limit is 10GiB, and forced_use is 11GiB, and THIS number is 90(%),
+ * we still consider 1GiB available for normal buffer usage. */
+static const size_t s_max_impact_of_forced_buffers_on_memory_limit_as_percentage = 80;
+
+struct aws_s3_default_buffer_pool {
+    struct aws_allocator *base_allocator;
+    struct aws_mutex mutex;
+
+    size_t block_size;
+    size_t chunk_size;
+    /* size at which allocations should go to secondary */
+    size_t primary_size_cutoff;
+
+    /* NOTE: See aws_s3_buffer_pool_usage_stats for descriptions of most fields */
+
+    size_t mem_limit;
+
+    size_t primary_allocated;
+    size_t primary_reserved;
+    size_t primary_used;
+
+    size_t secondary_reserved;
+    size_t secondary_used;
+
+    size_t forced_used;
+
+    struct aws_array_list blocks;
+
+    struct aws_linked_list pending_reserves;
+};
+
+struct s3_pending_reserve {
+    struct aws_linked_list_node node;
+    struct aws_future_s3_buffer_ticket *ticket_future;
+    struct aws_s3_buffer_pool_reserve_meta meta;
+};
+
+struct s3_buffer_pool_block {
+    size_t block_size;
+    uint8_t *block_ptr;
+    uint16_t alloc_bit_mask;
+};
+
+/*
+ * Sets n bits at position starting with LSB.
+ * Note: n must be at most 8, but in practice will always be at most 4.
+ * position + n should at most be 16
+ */
```

Comment:
The function name `s_default_ticket_claim` could be more descriptive. Consider `s_ticket_wrapper_claim_impl` to better indicate it's an implementation of the wrapper's claim function.